set_instance_tags/app.py

```diff
@@ -0,0 +1,127 @@
+import json
+import logging
+import re
+import traceback
+
+import boto3
+import requests
+
+from crhelper import CfnResource
+
+
+MISSING_INSTANCE_ID_ERROR_MESSAGE = 'InstanceId parameter is required'
+
+log = logging.getLogger(__name__)
+log.setLevel(logging.DEBUG)
+
+helper = CfnResource(
+  json_logging=False, log_level='DEBUG', boto_level='INFO')
+
+
+def get_ec2_client():
+  return boto3.client('ec2')
+
+
+def get_instance_id(event):
+  '''Get the instance id from event params sent to lambda'''
+  resource_properties = event.get('ResourceProperties')
+  instance_id = resource_properties.get('InstanceId')
+  if not instance_id:
+    raise ValueError(MISSING_INSTANCE_ID_ERROR_MESSAGE)
+  return instance_id
+
+
+def get_instance_tags(instance_id):
+  '''Look up the instance tags'''
+  client = get_ec2_client()
+  response = client.describe_tags(
+    Filters = [{'Name': 'resource-id', 'Values': [instance_id]}]
+  )
+  log.debug(f'EC2 describe tags response: {response}')
+  tags = response.get('Tags')
+  if not tags or len(tags) == 0:
+    raise Exception(f'No tags returned, received: {response}')
+
+  # format tags to easily pass to create_tags
+  for tag in tags:
+    tag.pop("ResourceId")
+    tag.pop("ResourceType")
+
+  return tags
+
+
+def get_principal_id(tags):
+  '''Find the value of the principal arn among the EC2 tags'''
+  principal_arn_tag = 'aws:servicecatalog:provisioningPrincipalArn'
+  for tag in tags:
+    if tag.get('Key') == principal_arn_tag:
+      principal_arn_value = tag.get('Value')
+      principal_id = principal_arn_value.split('/')[-1]
+      return principal_id
+  else:
+    raise ValueError('Could not derive a provisioningPrincipalArn from tags')
+
+
+def get_synapse_email(synapse_id):
+  '''Use the synapse id to look up the synapse user name'''
+  synapse_url = f'https://repo-prod.prod.sagebase.org/repo/v1/userProfile/{synapse_id}'
+  response = requests.get(synapse_url)
+  response.raise_for_status()
+  user_profile = response.json()
+  log.debug(f'Synapse user profile response: {user_profile}')
+  user_name = user_profile.get('userName')
+  synapse_email = f'{user_name}@synapse.org'
+  return synapse_email
+
+
+def get_owner_email(principal_id):
+  '''Derive an email from a id'''
+  email_check = re.compile(r"[^@]+@[^@]+\.[^@]+")
```

Comment:
from _an_ id

It's not clear to me why a 'principal_id' would turn out to be an email address.   Where does such an email address come from?

Reply:
the principal id is the synapse id which is used to retrieve the username and return  `<synapse user name>@synapse.org` as the email address.  isn't that the synapse email?

Reply:
When the original code was written, we had not completed implementing Synapse login, and were still using real IAM user accounts and manually setting OwnerEmail at least part of the time. At that time our principal ids mostly actually were email addresses. @zaro0508, up to you if you want to rewrite this function to assume that the principal ID will always be a Synapse ID now.

Reply:
The principal_id is a number, not an email address.  For example, yours is 3377358, @zaro0508 (see https://www.synapse.org/#!Profile:3377358)

Reply:
@brucehoff You're correct, that is the case _now_. It was not always the case. I wrote a comment above to provide some historical context for everyone.

Reply:
will remove email_check on a follow on PR.